pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+    This is a whitelist of file patterns which signal to the logic
+    of method `_frame_info_is_pyteal()` that a particular
+    frame was _NOT_ created by the user.
+
+    QUIRKS:
+    * Unfortunately, this means that if a user created file satisfies the above pattern,
+    the performance of the source mapper will degrade.
+    * User generated code that includes the pattern `NatalStackFrame` may produce degraded results
+    """
+
+    frame_info: FrameInfo
+    node: AST | None
+    creator: "NatalStackFrame"
+
+    # for debugging purposes:
+    full_stack: list[FrameInfo] | None = None
+
+    @classmethod
+    def _init_or_drop(
+        cls, creator: "NatalStackFrame", f: FrameInfo, full_stack: list[FrameInfo]
+    ) -> "StackFrame | None":
+        """
+        Attempt to create a StackFrame object.
+        However, if the resulting is considered "Python Crud" abandon and return None.
+        When debugging, also persist the full_stack that was provided.
+        """
+        node = cast(AST | None, Source.executing(f.frame).node)
+        frame = StackFrame(
+            f, node, creator, full_stack if NatalStackFrame._debug else None
+        )
+        return frame if frame._not_py_crud() else None
+
+    # TODO: when a source mapper is instantiated, it ought to survey
+    # the user's project files and warn in the case that some file
+    # matches the _internal_paths pattern
+    _internal_paths = [
+        "beaker/__init__.py",
+        "beaker/application.py",
+        "beaker/consts.py",
+        "beaker/decorators.py",
+        "beaker/state.py",
+        "pyteal/__init__.py",
+        "pyteal/ast",
+        "pyteal/compiler",
+        "pyteal/ir",
+        "pyteal/pragma",
+        "pyteal/stack_frame.py",
+        "tests/abi_roundtrip.py",
+        "tests/blackbox.py",
+        "tests/compile_asserts.py",
+        "tests/mock_version.py",
+    ]
+    _internal_paths_re = re.compile("|".join(_internal_paths))
+
+    def as_pyteal_frame(
+        self,
+        rel_paths: bool = True,
+        parent: "PyTealFrame | None" = None,
+    ) -> "PyTealFrame":
+        """
+        Downcast one level in the class hierarchy
+        """
+        return PyTealFrame(
+            frame_info=self.frame_info,
+            node=self.node,
+            creator=self.creator,
+            full_stack=self.full_stack,
+            rel_paths=rel_paths,
+            parent=parent,
+        )
+
+    @classmethod
+    def _frame_info_is_right_before_core(cls, f: FrameInfo) -> bool:
+        # We keep this method around to be used in a test . Originally,
+        # it was actually used in the __init__ method of StackFrame
+        # to calculate `last_drop_idx`. However, as -at the time of writing-
+        # last_drop_idx =ALWAYS= 1, the logic was simplified and
+        # this method was removed from the live calculation.
+        return bool(code := f.code_context or []) and "NatalStackFrame" in "".join(code)
```

Comment:
`...before_core` is intentional, as I mean the "core" or "heart" of the compilation. However, admittedly, this nomenclature is confusing. I anticipate of getting rid entirely of this method entirely in #676. However, just in case, I added a line item to #681:

<img width="436" alt="image" src="https://user-images.githubusercontent.com/291133/221313754-7618010d-eb16-4bd3-97fe-2999ddfb99fe.png">